main/modes/games/2048/mode_2048.c

```diff
@@ -0,0 +1,492 @@
+    .fnAudioCallback          = NULL,
+    .fnBackgroundDrawCallback = NULL,
+    .fnEspNowRecvCb           = NULL,
+    .fnEspNowSendCb           = NULL,
+    .fnAdvancedUSB            = NULL,
+};
+
+t48_t* t48;
+
+//==============================================================================
+// Functions
+//==============================================================================
+
+/**
+ * @brief Enter 2048 mode and set everything up
+ */
+static void t48EnterMode(void)
+{
+    setFrameRateUs(16667); // 60 FPS
+
+    // Init Mode & resources
+    t48 = calloc(sizeof(t48_t), 1);
+
+    // Load fonts
+    loadFont("ibm_vga8.font", &t48->font, false);
+    loadFont("sonic.font", &t48->titleFont, false);
+    makeOutlineFont(&t48->titleFont, &t48->titleFontOutline, false);
+
+    // Load images
+    t48->tiles = calloc(ARRAY_SIZE(tileSpriteNames), sizeof(wsg_t));
+    for (int32_t tIdx = 0; tIdx < ARRAY_SIZE(tileSpriteNames); tIdx++)
+    {
+        loadWsg(tileSpriteNames[tIdx], &t48->tiles[tIdx], true);
+    }
+
+    t48->sparkleSprites = calloc(ARRAY_SIZE(sparkleSpriteNames), sizeof(wsg_t));
+    for (int32_t sIdx = 0; sIdx < ARRAY_SIZE(sparkleSpriteNames); sIdx++)
+    {
+        loadWsg(sparkleSpriteNames[sIdx], &t48->sparkleSprites[sIdx], true);
+    }
+
+    t48->newSparkles = calloc(ARRAY_SIZE(newSparkleSprNames), sizeof(wsg_t));
+    for (int32_t sIdx = 0; sIdx < ARRAY_SIZE(newSparkleSprNames); sIdx++)
+    {
+        loadWsg(newSparkleSprNames[sIdx], &t48->newSparkles[sIdx], true);
+    }
+
+    // Load sounds
+    loadMidiFile("Follinesque.mid", &t48->bgm, true);
+    loadMidiFile("sndBounce.mid", &t48->click, true);
+
+    // Init Text Entry
+    textEntryInit(&t48->font, 4, t48->playerInitials);
+    textEntrySetBGColor(c001);
+    textEntrySetEmphasisColor(c500);
+    textEntrySetNewCapsStyle(true);
+    textEntrySetNewEnterStyle(true);
+    textEntrySetCapMode();
+
+    // Initialize the scores
+    t48InitHighScores();
+
+    // Initialize the game
+    t48->state = T48_START_SCREEN;
+    t48_initStartScreen(t48);
+}
+
+/**
+ * @brief Exit 2048 mode and free all resources
+ */
+static void t48ExitMode(void)
+{
+    freeFont(&t48->titleFontOutline);
+    freeFont(&t48->titleFont);
+    freeFont(&t48->font);
+
+    for (uint8_t i = 0; i < ARRAY_SIZE(sparkleSpriteNames); i++)
+    {
+        freeWsg(&t48->sparkleSprites[i]);
+    }
+    free(t48->sparkleSprites);
+
+    for (uint8_t i = 0; i < ARRAY_SIZE(tileSpriteNames); i++)
+    {
+        freeWsg(&t48->tiles[i]);
+    }
+    free(t48->tiles);
+
+    soundStop(true);
+    unloadMidiFile(&t48->bgm);
+    unloadMidiFile(&t48->click);
+
+    free(t48);
+}
+
+/**
+ * @brief The main game loop, responsible for input handling, game logic, and animation
+ *
+ * @param elapsedUs The time since this was last called, for animation
+ */
+static void t48MainLoop(int64_t elapsedUs)
+{
+    // Dim LEDs
+    t48_fadeLEDs(elapsedUs);
+
+    // Play BGM if it's not playing
```

Comment:
@dylwhich is there simple way to have a MIDI player loop automatically? Using the callback to restart it works, but it feels a little clunky.